Signup backend testing

PistolRcks opened the pull request:
Lots of additions along with the testing. Notably, Supertest is added. Need help with deleting artifacts after testing, so that would be greatly appreciated. Code coverage is up!

PistolRcks committed:
Separate insert user callback into separate function

PistolRcks committed:
Insert JSDocs

PistolRcks committed:
Start tests, put database init into own function

PistolRcks committed:
test *should* be done

just need to get the master branch pushed

PistolRcks committed:
single line looks nicer

PistolRcks committed:
Merge branch 'master' into signup_backend_testing

PistolRcks committed:
Merge branch 'master' into signup_backend_testing

PistolRcks committed:
Change password and username checking to throw errors

PistolRcks committed:
Fix up such that tests begin to work

PistolRcks committed:
Integrate username/password check

PistolRcks committed:
More tests...

PistolRcks committed:
fuckin assloads of shit

Lots more testing. Adding Supertest. Export changes. `insertUser` changed to async

PistolRcks commented:
fuck

PistolRcks committed:
Fix import issues with validators

`checkUsernameRequirements` and `checkPasswordRequirements` have been separated into their own files.

PistolRcks committed:
Merge branch 'master' into signup_backend_testing

PistolRcks committed:
Starting to work with `with-local-tmp-dir`

Will probably just change this to mock

PistolRcks committed:
Merge branch 'signup_backend_testing' of https://github.com/PistolRcks/web-puzzle-project into signup_backend_testing

Blackkirby72 commented on client/components/AccountCreation/AccountCreation.jsx:
```diff
@@ -22,13 +23,17 @@ function AccountCreation(props) {
+            }
+            else if(checkPasswordRequirements(formData.password) && checkUsernameRequirements(formData.username)) {
+                //TODO send the formData to the backend
+                console.log("Hey this code works");
```
can you remove this log while you're here? it should've been caught in user login front end review

PistolRcks replied:
Required for a test at here: 
https://github.com/PistolRcks/web-puzzle-project/blob/5f6151751a2ddc72e09c0ad664d49e38a503db96/tests/client/AccountCreation.test.js#L72 
Perhaps a better way of writing the test is in order? Either way, rewriting the test is out of scope. Resolving.

Blackkirby72 commented on server/api/api.js:
```diff
@@ -1,5 +1,5 @@
 const Express = require('express')
-const signup = require(`./signup`)
+const signup = require(`./signup`).signup
```
`const { signup } = require('./signup')`
^back ticks are formatted strings, which can decrease performance very slight

PistolRcks replied:
Resolved from 5f6151751a2ddc72e09c0ad664d49e38a503db96.

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,83 @@
+    return next(error.message);
+  }
+
+  var salt = Crypto.randomBytes(16);      // salt is required for hashing; it makes 
```
use `const` instead. `var` is outdated, and it's usually suggested to use `const` - short for "constant" and `let` - short for "let us not use var" (it's for a variable that will be changed at some point)

PistolRcks replied:
Resolved from 5f6151751a2ddc72e09c0ad664d49e38a503db96.

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,83 @@
\ No newline at end of file
+    // insert new user into database
+    try {
+      var newUser = await insertUser(db, req.body.username, hashedPassword, salt);
```
`const`

PistolRcks replied:
Resolved from 5f6151751a2ddc72e09c0ad664d49e38a503db96.

Blackkirby72 commented on tests/server/signup.test.js:
```diff
@@ -0,0 +1,63 @@
+import {describe, expect, test, beforeAll, afterAll, beforeEach} from '@jest/globals';
```
you shouldn't actually need to import these

PistolRcks replied:
Resolved from 5f6151751a2ddc72e09c0ad664d49e38a503db96.

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,83 @@
-var db = require('../db');
+const Crypto = require('crypto');
+const Sqlite3 = require('sqlite3');
+var db = require('../db').db;
```
`const { db } = require('../db')`

PistolRcks replied:
Resolved from 5f6151751a2ddc72e09c0ad664d49e38a503db96.

Blackkirby72 commented on server/db.js:
```diff
@@ -1,40 +1,43 @@
+  return db;
+};
+
+var db = initDatabase('./server/puzzle.db');
```
`const`

PistolRcks replied:
Resolved from 5f6151751a2ddc72e09c0ad664d49e38a503db96.

Blackkirby72 reviewed (commented):
The `signup` callback function should check if the username is currently in the database. Currently, when trying to signup an existing username, the server crashes due to a failing unique constraint

![image](https://user-images.githubusercontent.com/47089182/221951763-0df3c1fc-998b-4b76-a0dc-a810b01e557a.png)

Also, one way to deal with database modularization for unit testing purposes, is to have the callback function in `db.all()` to simply `return this` and deal with the result of the query in a separate function. Then you can mock `db.all()` to return data in the format you would expect. Otherwise, you could find a way to create a "test" database for your unit test suites, which would let you actually test that the database connection works as it should

Either method would be fine since the latter mostly tests that the database a) exists and b) runs the queries properly. Both of those should be caught in manual testing though

PistolRcks committed:
mock tests correctly now

filename change to better fit to standards, also tests are mocked correctly

current mocking solution is gross and should be reworked

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,83 @@
+    checkPasswordRequirements(req.body.password);
+  } catch (error) {
+    res.status(401).send(error.message);
+    return next(error.message);
```
i think returning `next()` makes it so the endpoint doesn't actually respond to the request. I'd just return `res.status...`

PistolRcks replied:
Resolved from 5f6151751a2ddc72e09c0ad664d49e38a503db96.

`return;` was used instead of `return next();` and the use of `next()` will be deprecated in the future in these circumstances.

PistolRcks committed:
Adjust to review specs

PistolRcks committed:
I think tests are complete now?

`handleUser` got changed to a callback because it's easier to test that way. Tests with callbacks now use `done` when possible as that waits until we're actually done to finish.

PistolRcks committed:
Prettification

Prettified all the files I modified in this PR

PistolRcks commented:
Code coverage is now at at least 80% for all files EXCEPT FOR `db.js`, which is being mocked.

(and also to be honest the testing of the database isn't part of my user story so I don't care)

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,102 @@
- * database.  If the record is successfully created, the user is logged in.
- * 
- * This was also shamelessly stolen from the passport tutorial.
+ * The `res` parameter should contain both `username` and `password` in its body.
```
it's the `req` param that has the data from the client

PistolRcks replied:
oops lol

PistolRcks replied:
Resolved in [89c391e](https://github.com/PistolRcks/web-puzzle-project/pull/24/commits/89c391e96c30d02d5b83bb1f35071283e25411ef).

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,102 @@
-      return next("Error: Username or password not set!");
-    }
+  // check that signup data is real
+  if (!req.body.username || !req.body.password) {
```
i don't think there'd be much of a performance impact due to how JS works, but you could destructure `username` and `password` at the beginning of the callback in order to have a shorter "name" for them

Not a huge deal though

PistolRcks replied:
Resolved in [89c391e](https://github.com/PistolRcks/web-puzzle-project/pull/24/commits/89c391e96c30d02d5b83bb1f35071283e25411ef).

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,102 @@
+      if (err) throw err;
+
+      // create a user object, handle it elsewhere
+      // TODO: Perhaps this should also generate a user token at this time?
```
assuming you meant session token: session token is created upon login. If we want sign up to automatically log in afterward, I can take care of that WPP-40 or Brett can in WPP-24

PistolRcks replied:
I did mean session token; and fair. Even at that point, the login process could probably be started here as well so that we don't have any unnecessary code duplication.

Blackkirby72 replied:
all it would take is requiring the `login` callback from `login.js` and calling it at the end, but we have to wait for Connor to finish up testing and merge it into main

PistolRcks replied:
Not gonna do it in this PR.

Blackkirby72 commented on tests/server/signupRoute.test.js:
```diff
@@ -0,0 +1,71 @@
+
+const noUNOrPass = "Error: Username or password not set!";
+
+describe("Test signup route", () => {
```
add 
```
beforeAll(() => {
    jest.spyOn(console, "log").mockImplementation(() => {});
    jest.spyOn(console, "error").mockImplementation(() => {});
  });
```
to stop the test from logging to the console

PistolRcks replied:
Resolved in [89c391e](https://github.com/PistolRcks/web-puzzle-project/pull/24/commits/89c391e96c30d02d5b83bb1f35071283e25411ef).

Blackkirby72 commented on tests/server/signup.test.js:
```diff
@@ -0,0 +1,86 @@
+import { insertUser } from "../../server/api/signup";
```
Personally, I would merge this test file with the other one. At the very lest, I'd name this one something like "signupDB.test.js" and the other "signupCallback.test.js"

PistolRcks replied:
Resolved in [89c391e](https://github.com/PistolRcks/web-puzzle-project/pull/24/commits/89c391e96c30d02d5b83bb1f35071283e25411ef). Files were merged.

Blackkirby72 commented on tests/server/signupRoute.test.js:
```diff
@@ -0,0 +1,71 @@
+
+  test("Response 200 - happy insertion", (done) => {
+    // this test *may* look a little gross, but this makes sure it hits all the lines of the test
+    request
+      .post("/api/signup")
+      .send({
+        username: "usernameVeryCool",
+        password: "Very1Epic",
+      })
+      .then((res) => {
+        expect(res.statusCode).toEqual(200);
+
+        db.get(
+          `select user_id from User where username="usernameVeryCool"`,
+          function (err, row) {
+            if (err) throw err;
+            expect(row["user_id"]).toBeDefined(); // we don't know the user_id; however, it should be something
+            done();
+          }
+        );
+      });
```
It could look a little cleaner by following the same `const res = request(...)` that you have in the other tests. Just pull the db call outside of the `.then()`

PistolRcks replied:
`done()` is required for callbacks to be handled correctly. You also can't use `done()` within an `async` context; you must either use `done()` in a non-async context or return a Promise (and I'm not sure how to handle that within a callback). So, it'll stay ugly.

cf. https://jestjs.io/docs/asynchronous

Blackkirby72 commented on server/api/signup.js:
```diff
@@ -1,51 +1,102 @@
+ * @returns Nothing.
+ */
+async function insertUser(db, username, hashedPassword, salt, callback) {
+  await db.run(
```
Before inserting the username + password, you should first check if there is an existing row with the same username. If there is, send an error and return from the method. If there isn't, then continue on and add the new user.

As is, the server (and app) crashes if you attempt to add a user with an existing username.
As an example, send a POST request to `/api/signup` with
```
{
    "username": "Test1234",
    "password": "Test1234"
}
```
twice, and the server will crash

*This comment is why I'm requesting changes since this should be handled before the user story can be considered complete.

PistolRcks replied:
Resolved in [079b59c](https://github.com/PistolRcks/web-puzzle-project/pull/24/commits/079b59c306c8712d40520e33c6b90f92a121b814). Callback within `insertUser` now returns an error object if an error is thrown by `db.run()`.

PistolRcks committed:
Consolidation of files, resolution of review comments

PistolRcks committed:
Minor changes, username constraint being caught now
